Updates following review #10 (#11)
buerokratt/Mock-Classifier: docs/technical/mock-classifier-technical-architecture.md

The Classifier will be hard-coded to support the following natural language resolutions for testing and demo purposes.

| Input Text (all in English)                                  | Calls back with |
| ------------------------------------------------------------ | --------------- |
| I want to register my child at school                        | education       |
| How do I file my annual tax information                      | economic        |
| I have a question about the Estonian pension system          | economic        |
| How do I get to Lahemaa park?                                | environment     |

> | How do I arrange for my COVID-19 booster vaccination         | social          |

Author: Actually no, the ministry which health comes under is https://en.wikipedia.org/wiki/Ministry_of_Social_Affairs_(Estonia)

## Token Resolutions

The Classifier will accept special tokens anywhere in the input text which will inform the Classifier which ministries to resolve.

The token format will be `{{token}}` where token is one of labels listed in the Ministries section.

Here are some examples:

- An input text of `Please return the {{social}} minsitry` will call back with `social`
- An input text of `{{environment}}` will call back with `environment`
- An input text of `{{rural}} please` will call back with `rural`

Each token detected will result in a separate call back to DMR. For example, input text of `I want to see {{rural}}{{social}} and {{environment}}` will return `rural`, `social` and `environment` as three separate DMR call backs.

We want the Classifier to support random ministry selection to help with testing, therefore is the `{{random}}` token is used, the Classifier will randomly select and return one of the 11 ministries. For example, an input text of `Give me a {{random}} one` will return a randomly selected ministry.

> Spaces are not permitted so an input text of `Please return {{ environment}}` would be treated as an error case.

Author: I actually think the opposite; having a single, well documented way to write tokens in messages will make it simpler to use. Flexibility in this case increases complexity and there is no need to have multiple ways to define a token.

> Spaces are not permitted so an input text of `Please return {{ environment}}` would be treated as an error case.

Author: Discussed offline with @decodingahmed and have agreed to keep it as it is.

> If a token is detected with unrecognised text (i.e. text that does not exactly map to one of the labels defined in the Ministries section or the word `random`), this will be treated as an error case. For example, the input text of `Please return {{educationandresearch}}` would be an error as would `{{Please return {{education,social}}`.

Reviewer: I think the `{{education,social}}` scenario is one that we should support also. It would save us from awkwardly typing }} {{ etc.   These scenarios are already written and tested in the POC.

Same as above, I'm open on this too but just want to suggest it so we leverage what's been done already instead of re-invent the wheel.

Author: My thoughts are the same as above on the spaces .... simplicity trumps flexibility and having a single, well-documented syntax for tokens will make it easier to use (most requests for developers are likely to come from saved curl, postman etc so will rarely be typed freehand). I appreciate that we already have the code for this in PoC but that should not determine the production approach.

Author: Discussed offline with @decodingahmed and have agreed to keep it as it is.

## API Design

> The Classifier will be a REST api with just a single endpoint which is `/classifications`.

Reviewer: Addressed in #12

#### Request

> Accepts only `POST` requests. Other http verbs will result in `404/NotFound` responses.

Reviewer: Addressed in #12

> Expects an `message` parameter which is an array of strings. The strings contains the text representing the input text. This will be defined in a JSON request body, for example

Reviewer: Addressed in #12

```json
{

> "message":["I want to register my child at school"]

Reviewer: Plural `messages` since it is an array.

```suggestion
"messages":["I want to register my child at school"]
```

}
```

or

```json
{

> "message":["Please return the {{social}} minsitry"]

Reviewer: Plural `messages`
```suggestion
"messages":["Please return the {{social}} ministry"]
```

}
```

> Other parameters may also be required in order for the Classifier to call back to DMR when the processing is completed (call-back uri for example). Please consult the DMR design documentation for details of these parameters.

Reviewer: Can I suggest that we spell "Callback" as a single word term really to stay consistent with industry standards.

```suggestion
Other parameters may also be required in order for the Classifier to callback to DMR when the processing is completed (callback URI for example). Please consult the DMR design documentation for details of these parameters.
```

#### Response

For the initial response, the Classifier will return `202/Accepted` with an empty body.

> The appropriate http status codes will be returned for error scenarios.

Reviewer: ```suggestion
The appropriate HTTP status codes will be returned for error scenarios.
```

> #### Call-back
> 
> The API will then do it's work to establish the ministry.
> 
> When the work is complete, the Classifier will call the DMR API with a POST request.
> 
> The request body will be json format following this schema (the value can be any of the label outlined in the Ministries section).

Reviewer: ```suggestion
#### Callback

The API will then do its work to establish the ministry.

When the work is complete, the Classifier will call the DMR API with a POST request.

The request body will be JSON format following this schema (the value can be any of the labels outlined in the Ministries section).
```
